motionplan/check.go

```diff
@@ -0,0 +1,430 @@
+	if err != nil {
+		return err
+	}
+	planEndPiF, ok := plan.Path()[len(plan.Path())-1][checkFrame.Name()]
+	if !ok {
+		return errors.New("check frame given not in plan Path map")
+	}
+	planEndPoseWorld, err := toWorld(planEndPiF, plan.Trajectory()[len(plan.Path())-1])
+	if err != nil {
+		return err
+	}
+
+	// setup the planOpts. Poses should be in world frame. This allows us to know e.g. which obstacles may ephemerally collide.
+	if sfPlanner.planOpts, err = sfPlanner.plannerSetupFromMoveRequest(
+		planStartPoseWorld.Pose(),
+		planEndPoseWorld.Pose(),
+		startingInputs,
+		worldState,
+		nil, // no pb.Constraints
+		nil, // no plannOpts
+	); err != nil {
+		return err
+	}
+
+	// create a list of segments to iterate through
+	segments := make([]*ik.Segment, 0, len(plan.Path())-wayPointIdx)
+	// get checkFrame's currentInputs
+	// *currently* it is guaranteed that a relative frame will constitute 100% of a solver frame's dof
+	checkFrameCurrentInputs, err := sf.mapToSlice(currentInputs)
+	if err != nil {
+		return err
+	}
+	arcEndInputs, err := sf.mapToSlice(plan.Trajectory()[wayPointIdx])
+	if err != nil {
+		return err
+	}
+
+	currentArcEndPose := spatialmath.Compose(expectedArcEndInWorld.Pose(), errorState)
+	// pre-pend to segments so we can connect to the input we have not finished actuating yet
+	segments = append(segments, &ik.Segment{
+		StartPosition:      currentPoseInWorld.Pose(),
+		EndPosition:        currentArcEndPose,
+		StartConfiguration: checkFrameCurrentInputs,
+		EndConfiguration:   arcEndInputs,
+		Frame:              sf,
+	})
+
+	lastArcEndPose := currentArcEndPose
+	// iterate through remaining plan and append remaining segments to check
+	for i := wayPointIdx + 1; i <= len(plan.Path())-1; i++ {
+		thisArcEndPoseTf, ok := plan.Path()[i][checkFrame.Name()]
+		if !ok {
+			return errors.New("check frame given not in plan Path map")
+		}
+		thisArcEndPoseInWorld, err := toWorld(thisArcEndPoseTf, plan.Trajectory()[i])
+		if err != nil {
+			return err
+		}
+		thisArcEndPose := spatialmath.Compose(thisArcEndPoseInWorld.Pose(), errorState)
+		// Starting inputs for relative frames should be all-zero
+		startInputs := map[string][]referenceframe.Input{}
+		for k, v := range plan.Trajectory()[i] {
+			if k == checkFrame.Name() {
+				startInputs[k] = make([]referenceframe.Input, len(checkFrame.DoF()))
+			} else {
+				startInputs[k] = v
+			}
+		}
+		segment, err := createSegment(sf, lastArcEndPose, thisArcEndPose, startInputs, plan.Trajectory()[i])
+		if err != nil {
+			return err
+		}
+		lastArcEndPose = thisArcEndPose
+		segments = append(segments, segment)
+	}
+
+	// go through segments and check that we satisfy constraints
+	// TODO(RSDK-5007): If we can make interpolate a method on Frame the need to write this out will be lessened and we should be
+	// able to call CheckStateConstraintsAcrossSegment directly.
+	var totalTravelDistanceMM float64
+	for _, segment := range segments {
```

Comment:
without going line by line (so I may be wrong) I think this for-loop is almost identical to the for loop within `checkPlanAbsolute`. We can consider moving this for - loop if this is true into a separate function for better readability.